Validation/EventGenerator/interface/BasicHepMCValidation.h

```diff
@@ -43,77 +45,148 @@ class BasicHepMCValidation : public DQMEDAnalyzer{
-	//
-	MonitorElement *piplusMomentum, *piminusMomentum, *pizeroMomentum, *KplusMomentum, *KminusMomentum, *KlzeroMomentum,  *KszeroMomentum;
-	//
-	MonitorElement *pMomentum, *pbarMomentum, *nMomentum, *nbarMomentum, *l0Momentum, *l0barMomentum;
-	//
-	MonitorElement *DplusMomentum, *DminusMomentum, *DzeroMomentum,  *BplusMomentum, *BminusMomentum, *BzeroMomentum, *BszeroMomentum;
-	//
-	MonitorElement *otherPtclMomentum;
-
-	///other ME's
-	MonitorElement *genPtclNumber; 
-	MonitorElement *genVrtxNumber;
-	MonitorElement *unknownPDTNumber;
-	MonitorElement *outVrtxPtclNumber;
+	p_final = dqm.book1dHisto(name+"MomentumFinal", "log_{10}(P^{final}) of "+name+"s at end of decay chain", 
+				  60, -2, 4,"log_{10}(P^{final}) (log_{10}(GeV))","Number of "+name);
+	
+	lifetime_final=dqm.book1dHisto(name+"LifeTimeFinal", "Log_{10}(life-time^{final}) of "+name+"s at end of decay chain",
+					100,-15,-5,"Log_{10}(life-time^{final}) (log_{10}(s))","Number of "+name);
+      }
+
+      ~ParticleMonitor(){};
+      
+      bool Fill(const HepMC::GenParticle* p, double weight){
+	if(p->pdg_id()==pdgid){
+	  if(isFirst(p)){
+	    p_init->Fill(log10(p->momentum().rho()),weight);
+	    eta_init->Fill(p->momentum().eta(),weight);
+            const HepMC::GenParticle* pf=GetFinal(p); // inlcude mixing
+            p_final->Fill(log10(pf->momentum().rho()),weight);
+	    // compute lifetime...
+	    if(p->production_vertex() && p->end_vertex()){
+	      TVector3 PV(p->production_vertex()->point3d().x(),p->production_vertex()->point3d().y(),p->production_vertex()->point3d().z()); 
+	      TVector3 SV(p->end_vertex()->point3d().x(),p->end_vertex()->point3d().y(),p->end_vertex()->point3d().z()); 
+	      TVector3 DL=SV-PV; 
+	      double c(2.99792458E8),Ltau(DL.Mag()/100)/*cm->m*/,beta(p->momentum().rho()/p->momentum().m()); 
+	      double lt=Ltau/(c*beta);
+	      if(lt>1E-16)lifetime_init->Fill(log10(lt),weight);
+	      if(pf->end_vertex()){
+		TVector3 SVf(pf->end_vertex()->point3d().x(),pf->end_vertex()->point3d().y(),pf->end_vertex()->point3d().z());
+		DL=SVf-PV;
+		Ltau=DL.Mag()/100;
+		lt=Ltau/(c*beta);
+		if(lt>1E-16)lifetime_final->Fill(log10(lt),weight);
+	      }
+	    }
+	    count++;
+	  }
+	  return true;
+	}
+	return false;
+      }
+      
+      void FillCount(double weight){
+	if(nlog) numberPerEvent->Fill(log10(count),weight);
+	else numberPerEvent->Fill(count,weight);
+	count=0;
+      }
+      
+      int PDGID(){return pdgid;}
+      
+    private:
+      bool isFirst(const HepMC::GenParticle* p){
+	if(p->production_vertex()){
+	  for(HepMC::GenVertex::particles_in_const_iterator m=p->production_vertex()->particles_in_const_begin(); m!=p->production_vertex()->particles_in_const_end();m++){
+	    if(abs((*m)->pdg_id())==abs(p->pdg_id())) return false;
+	  }
+	}
+	return true;
+      }
+      
+      const HepMC::GenParticle* GetFinal(const HepMC::GenParticle* p){ // includes mixing
```

Comment:
@davidlange6 , this is PR after which couple of workflows (533.0 and 534.0) in 76X are not finishing (timing out after 1 hour). Looks like  GetFinal method is getting called recursively